0000-stream/0000-stream.md

```diff
@@ -0,0 +1,407 @@
+
+### 3.5. Packets and Frames
+
+STREAM packets are encoded, [encrypted](#encryption), and sent as the `data` field in [ILP Prepare, Fulfill, or Reject packets](../0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-prepare).
+
+Each STREAM packet consists of multiple [frames](#53-frames), which can be used to send money, data, or control-related information.
+
+### 3.6. Packet Acknowledgements (ACKs)
+
+STREAM uses ILP Fulfill packets as acknowledgements of the frames contained in the corresponding Prepare packet. If a receiver wishes to reject or NACK one or more frames in a given Prepare packet, they MUST respond with an ILP Reject packet.
+
+## 4. Life of a Connection
+
+This section describes how connections and streams are created, used, and closed.
+
+### 4.1. Setup
+
+A server MUST communicate the following values to a client using an **authenticated, encrypted** communication channel (such as HTTPS). Key exchange is NOT provided by STREAM.
+
+- STREAM Version (optional -- assumed to be version 1 unless specified)
+- Server ILP Address
+- Cryptographically secure random or pseudorandom shared secret (it is RECOMMNEDED to use 32 bytes)
+
+To avoid storing a 32 byte secret for each connection, a server MAY deterministically generate the shared secret for each connection from a single server secret and a nonce appended to the ILP Address given to a particular client, for example by using an HMAC.
+
+### 4.2. Matching Packets to Connections
+
+Incoming packets can either be associated with an existing connection, or, for servers, potentially create a new connection. Endpoints MAY append extra segments to the ILP addresses assigned to them by their upstream connectors to help direct incoming packets.
+
+STREAM packets are completely encrypted so endpoints must try to decrypt and parse them to determine whether a given packet was sent by the other endpoint of a connection. Incoming Prepare packets whose data cannot be decrypted with the expected shared secret MUST be rejected with `F06: Unexpected Payment` errors.
+
+### 4.3. Client Address Communication and Connection Migration
+
+When a client connects to a server, they MUST communicate their ILP address to the server using a `ConnectionNewAddress` frame.
+
+Either endpoint MAY change their ILP address at any point during a connection by sending a `ConnectionNewAddress` frame.
+
+An attacker could cause an unsuspecting endpoint to participate in a Denial of Service (DoS) attack on a third-party endpoint. To mitigate this, endpoints SHOULD refrain from sending large numbers of packets or large amounts of data to a new ILP address before validating that the ILP address is controlled by the same party that knows the shared secret. STREAM uses the authenticated request/response packets in lieu of [QUIC's explicit Path Validation](https://quicwg.github.io/base-drafts/draft-ietf-quic-transport.html#rfc.section.6.7).
```

Comment:
Doesn't the authenticated encryption make it impossible for an attacker to pull this attack off? This paragraph makes it sound like that attack is possible under the current protocol.

Reply:
Authenticated encryption means you know if you're not talking to the right party. However, the point of this attack is to create a legitimate connection to you and then tell you I've changed my address to that of the victim so that you bombard the new address with data. The mitigation for this is that if you get an address change you probably shouldn't send more than one packet before getting a valid (authenticated) message back from the new address.